### PR TITLE
DrivAerML: EMA=0.999 + gc=0.5 compound at champion architecture

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

EMA alone failed at 9.749% on DrivAerML. gc alone at 0.5 was 4.54% (worse than 3.997%). But EMA + gc compound might work synergistically: gc prevents gradient spikes that destabilize training, EMA smooths the resulting checkpoint sequence. Testing this specific combination.

## Instructions

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.999 --grad-clip 0.5 \
  --wandb_group dm-ema-sweep
```

## Reporting Requirements
- Best val `surface_rel_l2_pct` (EMA model)
- Test at best val checkpoint
- W&B run ID

## Baseline

**DrivAerML baselines:**
- val best: `3.997%` @ ep467 (PR #2898, W&B: `bht6h42t`, 4L/512d, AdamW lr=5e-4, T_max=30)
- test best: `6.244%` (PR #2648, 4L/320d, truncated eval — OLDER CONFIG)
- External targets: AB-UPT 3.82% | Transolver-3: 3.71%

Note: EMA alone = 9.749% (#2899), gc=0.5 alone = 4.54% (#3010). Need to beat 3.997%.

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT faye:

## Results

### EMA=0.999 + gc=0.5 compound — NEGATIVE (diverged)

| Metric | Value | Baseline | Delta |
|--------|-------|----------|-------|
| Best val `surface_rel_l2_pct` | **9.953%** (ep113) | 3.997% (ep467) | +149% worse |
| Test at best val | N/A (diverged before completion) | 6.244% | — |
| External targets | AB-UPT 3.82% / Transolver-3 3.71% | — | — |

**W&B run:** `x3wvqyqa` (wandb-applied-ai-team/senpai-v1)

**Command used:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.999 --grad-clip 0.5 \
  --agent faye --wandb-name "faye/dm-ema-999-gc05-compound" \
  --wandb-group dm-ema-sweep
```

**Peak memory:** Not tracked by train.py (single GPU, batch-size 1, 4L/512d — consistent with prior DrivAerML runs)

### What happened

The EMA+gc compound hypothesis **did not work**. The model trained normally through epoch 113 (reaching best val 9.953%), then suffered catastrophic gradient explosion:

- **Epoch 111:** grad_norm_mean = 15.8 (first sign of instability)
- **Epoch 117:** grad_norm_mean = 116 (explosion begins)
- **Epoch 133:** grad_norm_mean = Infinity (full divergence)
- **Epochs 133-200:** Stuck at ~72% val with Infinity grad norms. Killed at epoch 200.

The divergence occurred at a cosine LR peak — gc=0.5 was insufficient to prevent the gradient spike that destabilized EMA-weighted training. The synergy hypothesis (gc prevents spikes → EMA smooths the sequence) failed because gc=0.5 is too weak to contain the DrivAerML gradient landscape at lr=5e-4.

**Key comparisons:**
- EMA alone (#2899): 9.749% — slightly better than EMA+gc (9.953%)
- gc=0.5 alone (#3010): 4.54% — much better without EMA
- No EMA, no gc (#2898): 3.997% — current champion

EMA consistently hurts DrivAerML regardless of gradient clipping. The no-EMA baseline at 3.997% remains the best approach.

### Suggested follow-ups

1. **EMA is harmful for DrivAerML** — all three EMA experiments (alone, +gc=0.5, decay=0.9999) perform 2-3x worse than no-EMA. Consider marking EMA as contraindicated for this dataset.
2. **If EMA must be tested further**, try much stronger gc (e.g. gc=0.1 or gc=0.2) to actually contain the gradient spikes at cosine peaks, though the fundamental issue may be that EMA averaging hurts the diverse surface-point geometry.
3. **Focus DrivAerML efforts on no-EMA configs** — the 3.997% → 3.71% gap to external targets is more likely closed by architecture/loss changes than by EMA smoothing.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3073 because ADVISOR: Closing — EMA=0.999 + gc=0.5 compound is a dead end on DrivAerML. Best val 9.953% at ep113 (+149% vs 3.997% baseline), crashed at ep191. gc=0.5 was insufficient — gradient spikes of 293 and 153 broke through the clip threshold before terminal divergence at ep136. EMA has now been tested in three configurations on DM (alone=9.749%, +gc=9.953%, +gc+WD=crash) — all dramatically worse than no-EMA champion. EMA is definitively contraindicated for DrivAerML. Good thorough analysis, faye..
